edb/pgsql/compiler/astutils.py

```diff
@@ -139,84 +139,113 @@ def array_get_inner_array(
+        return pgast.SelectStmt(
+            target_list=[
+                pgast.ResTarget(val=pgast.ColumnRef(name=['f1'])),
+            ],
+            from_clause=[
+                pgast.RangeFunction(
+                    functions=[
+                        pgast.FuncCall(
+                            name=('unnest',),
+                            args=[
+                                pgast.ArrayExpr(
+                                    elements=[wrapped_array],
                                 )
-                            )
-                        ]
-                    )
-                ]
+                            ],
+                        )
+                    ]
+                )
+            ]
+        )
+    else:
+        # temporary array of array implemented as array<tuple<array<...>>>
+        # using unnamed tuple
+        """
+        Postgres does not support arbitrarily accessing fields out of unnamed
+        composites and so we need to do an extra unnest(array[]) to be able to
+        specify the name and type our resulting array.
+
+        For example, the query: `select [[1]][0];` will produce the following
+        SQL:
+
+        SELECT
+                "expr-6~2"."array_value~4" AS "array_serialized~1"
+            FROM
+                LATERAL
+                (SELECT
+                        "expr-5~2"."array_value~3" AS "array_value~4"
+                    FROM
+                        LATERAL
+                        (SELECT
+                                (SELECT
+                                        "0"
+                                    FROM
+                                        -- EXTRA unnest(array[])
+                                        unnest(ARRAY[
+                                            -- INDEX INDIRECTION
+                                            edgedb_v7_2f26206480._index(
+                                                "expr-3~2"."array_value~2",
+                                                ($2)::int8,
+                                                'ERROR MESSAGE'
+                                            )
+                                        ]) AS ("0" int8[])
+                                ) AS "array_value~3"
+                            FROM
+                                LATERAL
+                                -- INITAL ARRAY [[1]]
+                                (SELECT
+                                        ARRAY[ROW("expr-2~2"."array_value~1")]
+                                        AS "array_value~2"
+                                    FROM
+                                        LATERAL
+                                        (SELECT
+                                                ARRAY[($1)::int8]
+                                                AS "array_value~1"
+                                        ) AS "expr-2~2"
+                                ) AS "expr-3~2"
+                        ) AS "expr-5~2"
```

Comment:
I know this is from a previous PR, but trim the comment example code down. It includes some extra irrelevant stuff: an extra wrapper to produce serialized, the toplevel NULL filter, an inject LIMIT